homework_1/02_ProgressiveTax/taxes/taxes.go

```diff
@@ -0,0 +1,60 @@
+package taxes
+
+import "github.com/pkg/errors"
+
+type taxBrackets struct {
+	brackets []float32
+	taxes    []float32
+}
+
+func checkTaxBrackets(brackets []float32, taxes []float32) error {
+	if len(brackets) == 0 || brackets[0] != 0 {
+		return errors.New("First bracket is not zero.")
+	}
+
+	if len(brackets) != len(taxes) {
+		return errors.New("Brackets and taxes don't have the same length.")
+	}
+
+	for idx, bracket := range brackets {
+		if idx > 0 && bracket <= brackets[idx - 1] {
+			return errors.New("Bracket is not greater than the previous.")
+		}
+
+		if taxes[idx] < 0 {
+			return errors.New("Tax is negative.")
+		}
+	}
+
+	return nil
+}
+
+func NewTaxBrackets(brackets []float32, taxes []float32) (*taxBrackets, error) {
+	if err := checkTaxBrackets(brackets, taxes); err != nil {
+		return nil, err
+	}
+
+	return &taxBrackets{brackets: brackets, taxes: taxes}, nil
+}
+
+func CalculateProgressiveTax(value float32, taxBrackets *taxBrackets) (float32, error) {
+	if value < 0 {
+		return 0, errors.New("Value is negative.")
+	}
+
+	brackets := taxBrackets.brackets
+	taxes := taxBrackets.taxes
+
+	var finalTax float32
```

Comment:
Even though Go initialises variable as zero-values it would be better to explicitly initialise this to zero here.
Something like:
* `finalTax := 0.0` (this would make it `float64` though) or
* `var finalTax float32 = 0.0`

You express your intent more clearly that way.